src/Module/RenderConfirmationBookingsHandler.php

```diff
@@ -0,0 +1,279 @@
+        $this->_setFallbackTz($fallbackTz);
+        $this->bookingRowTemplate = $bookingRowTemplate;
+        $this->bookingsSelectRm   = $bookingsSelectRm;
+        $this->servicesSelectRm   = $servicesSelectRm;
+        $this->exprBuilder        = $exprBuilder;
+        $this->bookingFormat      = $bookingFormat;
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * @since [*next-version*]
+     */
+    public function __invoke()
+    {
+        $event = func_get_arg(0);
+
+        if (!($event instanceof EventInterface)) {
+            throw $this->_createInvalidArgumentException(
+                $this->__('Argument is not an event instance'), null, null, $event
+            );
+        }
+
+        $payment   = $event->getParam(0);
+        $paymentId = $payment->ID;
+
+        // Alias expression builder
+        $b = $this->exprBuilder;
+
+        // Fetch the corresponding booking from storage
+        $condition = $b->eq($b->var('payment_id'), $b->lit($paymentId));
+        $bookings  = $this->bookingsSelectRm->select($condition);
+
+        $rows = '';
+        foreach ($bookings as $_booking) {
+            if (!($_booking instanceof BookingInterface)) {
+                throw $this->_createOutOfRangeException(
+                    $this->__('Booking is not a valid booking instance'), null, null, $_booking
+                );
+            }
+
+            $rows = $this->_renderBookingRow($_booking);
+        }
+
+        echo $this->_getTemplate()->render([
+            'table_heading'        => $this->__('Bookings'),
+            'service_column'       => $this->__('Service'),
+            'booking_start_column' => $this->__('Start'),
+            'booking_end_column'   => $this->__('End'),
+            'booking_tz_column'    => $this->__('Timezone'),
+            'booking_rows'         => $rows,
+        ]);
+    }
+
+    /**
+     * Renders a booking row.
+     *
+     * @since [*next-version*]
+     *
+     * @param BookingInterface $booking The booking instance.
+     *
+     * @return string|Stringable The rendered booking row.
+     */
+    protected function _renderBookingRow(BookingInterface $booking)
+    {
+        $timezone = $this->_getDisplayTimezone($booking);
+
+        // Get timestamps from booking
+        $startTs = $booking->getStart();
+        $endTs   = $booking->getEnd();
+
+        // Create date time helper instances
+        $startDt = Carbon::createFromTimestampUTC($startTs);
+        $endDt   = Carbon::createFromTimestampUTC($endTs);
+
+        // Shift to client timezone, if available
+        if ($timezone !== null) {
+            $startDt->setTimezone($timezone);
+            $endDt->setTimezone($timezone);
+        }
+
+        // Format times to strings
+        $startStr = $startDt->format($this->bookingFormat);
+        $endStr   = $endDt->format($this->bookingFormat);
+
+        return $this->bookingRowTemplate->render([
+            'service_name' => $this->_getServiceName($booking),
```

Comment:
Looks like your code makes an amount of DB queries equal to the number of bookings. There seems to be a good chance that multiple bookings will be for the same service. Is it feasible to optimize this right now?

Reply:
Good point. I'll see if I can make the service query only once.

Reply:
OK I misunderstood. The bookings are retrieved for a payment ID. Therefore, there _is_ a chance that multiple bookings will be for the same service, but it's equal to the chance of them _not_ being for the same service 😀 

@XedinUnknown How do you propose I optimize this?

Reply:
If the chance of them being for the same service is the same as *not* being for the same service, that means its 50%, which is high enough to optimize this.

As for *how* to optimize: I don't know. Maybe, use a cache. How long would it take you to plugin in a memoizer?

Reply:
Depends on how long composer bitches about SSL issues.